logging/ratelimit.go

```diff
@@ -0,0 +1,79 @@
+package logging
+
+import "golang.org/x/time/rate"
+
+type rateLimitedLogger struct {
+	next    Interface
+	limiter *rate.Limiter
+}
+
+// NewRateLimitedLogger returns a logger.Interface that is limited to a number
+// of logs per second
+func NewRateLimitedLogger(logger Interface, logsPerSecond rate.Limit) Interface {
+	return &rateLimitedLogger{
+		next:    logger,
+		limiter: rate.NewLimiter(logsPerSecond, 1),
+	}
+}
+
+func (l *rateLimitedLogger) Debugf(format string, args ...interface{}) {
+	if l.limiter.Allow() {
+		l.next.Debugf(format, args...)
+	}
+}
+
+func (l *rateLimitedLogger) Debugln(args ...interface{}) {
+	if l.limiter.Allow() {
+		l.next.Debugln(args...)
+	}
+}
+
+func (l *rateLimitedLogger) Infof(format string, args ...interface{}) {
+	if l.limiter.Allow() {
+		l.next.Infof(format, args...)
+	}
+}
+
+func (l *rateLimitedLogger) Infoln(args ...interface{}) {
+	if l.limiter.Allow() {
+		l.next.Infoln(args...)
+	}
+}
+
+func (l *rateLimitedLogger) Errorf(format string, args ...interface{}) {
+	if l.limiter.Allow() {
+		l.next.Errorf(format, args...)
+	}
+}
+
+func (l *rateLimitedLogger) Errorln(args ...interface{}) {
+	if l.limiter.Allow() {
+		l.next.Errorln(args...)
+	}
+}
+
+func (l *rateLimitedLogger) Warnf(format string, args ...interface{}) {
+	if l.limiter.Allow() {
+		l.next.Warnf(format, args...)
+	}
+}
+
+func (l *rateLimitedLogger) Warnln(args ...interface{}) {
+	if l.limiter.Allow() {
+		l.next.Warnln(args...)
+	}
+}
+
+func (l *rateLimitedLogger) WithField(key string, value interface{}) Interface {
+	return &rateLimitedLogger{
+		next:    l.next.WithField(key, value),
+		limiter: rate.NewLimiter(l.limiter.Limit(), 0),
+	}
+}
+
+func (l *rateLimitedLogger) WithFields(f Fields) Interface {
+	return &rateLimitedLogger{
+		next:    l.next.WithFields(f),
+		limiter: rate.NewLimiter(l.limiter.Limit(), 0),
```

Comment:
Why do we get a new limiter here?
Also, does it work with a burst size of zero?

Reply:
Yes, the burst limit was wrong.  Added a test and fixed.  As far as the new limiter I figured calling `WithFields()` created a new logger that should have an independent rate limit, but I can go either way on this.

Reply:
Thing is, code often calls `WithFields()` on every operation, e.g. the example you gave in the description, or this one:
https://github.com/weaveworks/common/blob/61ffdd448099f94ec3dc9aaf93705399b97a2bb2/middleware/grpc_logging.go#L30

I expect the case you're trying to hit is where a high volume of logs from from lots of operations, so you would want the same rate-limit to be applied across them all.  Try it!

Reply:
You're right.  Thanks man :+1: 

Fixed.